File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> result; vector<int> current;
        finalCombinations(0,target,candidates,current,result);
        return result;
    }
    void finalCombinations(int index,int target,vector<int>& candidates,vector<int>& current,
                          vector<vector<int>>& result){
        if(index==candidates.size()){
            if(target==0) result.push_back(current);
            return;
        }
        if(candidates[index]<=target){
            current.push_back(candidates[index]);
            finalCombinations(index,target-candidates[index],candidates,current,result);
            current.pop_back();
        }
        finalCombinations(index+1,target,candidates,current,result);
    }
};

File: 39-combination-sum/39-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result=new ArrayList<>();
        if(candidates==null || candidates.length==0) return result;
        backtrack(result,candidates,target,new ArrayList<>(),0);
        return result;
    }
    private void backtrack(List<List<Integer>> result,int[] candidates,int target,List<Integer> current,int index){
        if(index==candidates.length){
            if(target==0) result.add(new ArrayList<>(current));
            return;
        }
        if(candidates[index]<=target){
            current.add(candidates[index]);
            backtrack(result,candidates,target-candidates[index],current,index);
            current.remove(current.size()-1);
        }
        backtrack(result,candidates,target,current,index+1);
    }
}